### PR TITLE
Update trial old deps CI to use poetry 1.2.0

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

**Review by reivilibre — APPROVED** [review]
...do we really have another poetry pin in here? :rofl:

**Comment by erikjohnston** [comment]
I've substantially rewritten this to make it a) work, b) actually use py3.7, and c) cache stuff.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -150,16 +150,42 @@ jobs:
...
+      # pyproject.toml / poetry.lock.
+      - uses: actions/cache@v3
+        id: cache-poetry-old-deps
+        name: Cache poetry.lock
         with:
-          workdir: /github/workspace
-          entrypoint: .ci/scripts/test_old_deps.sh
+          path: |
+            poetry.lock
+            pyproject.toml
```
Comment (reivilibre) [inline .github/workflows/tests.yml:175]: I found this confusing; maybe worth a comment noting that the reason we cache pyproject.toml is because the step beneath patches it.

**Review by reivilibre — APPROVED** [review]
Seems sensible to me

**Review by DMRobertson — CHANGES_REQUESTED** [review]
There's a TODO on lines +26 -- +30 which has been done (see lines +20 and +34) but I forgot to remove. Can you remove the TODO while we're here?

File: .ci/scripts/prepare_old_deps.sh [hunk 2]
```diff
@@ -69,15 +59,12 @@ with open('pyproject.toml', 'w') as f:
 "
 python3 -c "$REMOVE_DEV_DEPENDENCIES"
 
-pipx install poetry==1.1.14
-~/.local/bin/poetry lock
+pip install poetry==1.2.0
+poetry lock
```
Comment (DMRobertson) [inline .ci/scripts/prepare_old_deps.sh:57]: FWIW it's possible that poetry's dependencies might conflict with those of `toml` and `wheel` here. It's probably not a problem but calling it out for paranoia's sake.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (DMRobertson) [reply .github/workflows/tests.yml:177]: My understanding:

- We compute a cache key based on the pyproject.toml in the checkout
- If there's a hit, we retrieve the _patched_ pyproject.toml and poetry.lock from the cache
- Otherwise we patch pyproject.toml and regenerate the lockfile
- During GHA cleanup we will write the patched pyproject.toml and poetry.lock to the cache for future runs

It does need a little bit of thought though: a quick comment SoundsGTM

**Review by DMRobertson — APPROVED** [review]

File: .ci/scripts/prepare_old_deps.sh (hunk 2 above)
Comment (erikjohnston) [reply .ci/scripts/prepare_old_deps.sh:57]: Shouldn't `poetry lock` here be ignoring toml and wheels deps?

Comment (DMRobertson) [reply .ci/scripts/prepare_old_deps.sh:57]: You're right that the virtualenv managed by poetry will be independent of toml and wheel. What I mean: the depedencies of the poetry package itself, i.e. the stuff listed here

```
$ pipx runpip poetry show poetry
Name: poetry
Version: 1.2.0
Summary: Python dependency management and packaging made easy.
Home-page: https://python-poetry.org/
Author: Sébastien Eustace
Author-email: sebastien@eustace.io
License: MIT
Location: /home/dmr/.local/pipx/venvs/poetry/lib/python3.10/site-packages
Requires: cachecontrol, cachy, cleo, crashtest, dulwich, html5lib, jsonschema, keyring, packaging, pexpect, pkginfo, platformdirs, poetry-core, poetry-plugin-export, requests, requests-toolbelt, shellingham, tomlkit, urllib3, virtualenv
Required-by: poetry-plugin-export
```

might have conflicts with the toml and wheel versions we've just installed.

But a) it doesn't look like there are any conflicts, and b) I think pip would at least warn us about them.

TL;DR ignore this comment, it's all good
